File: agent/ai-platform/frontend/src/types/skill.ts
/**
 * Skill-related type definitions for the frontend.
 *
 * Mirrors backend's src/skills/models.py — Skill, SkillCreateRequest,
 * SkillUpdateRequest, SkillListResponse — converted to camelCase TS.
 */

// ===== Enums =====

/** Origin of a Skill. */
export type SkillSource = "custom" | "mcp" | "builtin";

/** Lifecycle status of a Skill. */
export type SkillStatus = "active" | "inactive" | "deprecated";

/** Top-level Skill categories aligned with business systems. */
export type SkillCategory =
  | "finance"
  | "retail"
  | "department_store"
  | "hr"
  | "property"
  | "crm"
  | "valuecard"
  | "built_in";

// ===== Skill Parameter =====

/** A single parameter definition for a Skill (JSON-Schema-like). */
export interface SkillParameter {
  name: string;
  type: string;
  description: string;
  required: boolean;
  default: unknown;
  enum: unknown[] | null;
}

// ===== Skill =====

/** Canonical Skill representation used throughout the platform. */
export interface Skill {
  skillId: string;
  name: string;
  description: string;
  category: SkillCategory;
  tags: string[];
  parameters: Record<string, unknown>;
  requiredPermissions: string[];
  handler: string;
  timeout: number;
  version: string;
  status: SkillStatus;
  source: SkillSource;
  priority: number;
  requiresApproval: boolean;
  mcpServer: string | null;
  callCount: number;
  lastCalledAt: string | null;
}

// ===== Skill Score (for ranking display) =====

/** A Skill paired with its retrieval/ranking score. */
export interface SkillScore {
  skill: Skill;
  score: number;
  semanticSimilarity: number;
  usageFrequency: number;
  recencyBonus: number;
  categoryMatch: number;
}

// ===== Create / Update Requests =====

/** Request body for creating a new Skill via API. */
export interface SkillCreateRequest {
  skillId: string;
  name: string;
  description: string;
  category: SkillCategory;
  tags: string[];
  parameters: Record<string, unknown>;
  requiredPermissions: string[];
  handler: string;
  timeout: number;
  version: string;
  priority: number;
  requiresApproval: boolean;
}

/** Request body for partially updating a Skill. */
export interface SkillUpdateRequest {
  name?: string;
  description?: string;
  category?: SkillCategory;
  tags?: string[];
  parameters?: Record<string, unknown>;
  requiredPermissions?: string[];
  handler?: string;
  timeout?: number;
  version?: string;
  status?: SkillStatus;
  priority?: number;
  requiresApproval?: boolean;
}

// ===== List Response =====

/** Paginated list of Skills. */
export interface SkillListResponse {
  items: Skill[];
  total: number;
  page: number;
  pageSize: number;
}

// ===== Statistics =====

/** Registry statistics response. */
export interface SkillStats {
  total: number;
  active: number;
  inactive: number;
  byCategory: Record<string, number>;
  bySource: Record<string, number>;
}

File: agent/ai-platform/backend/src/models/skill.py
"""Skill 持久化的 SQLAlchemy ORM 模型。"""

from __future__ import annotations

from sqlalchemy import Boolean, DateTime, Float, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from src.models.base import Base, SoftDeleteMixin, TimestampMixin, UUIDPrimaryKeyMixin


class SkillModel(Base, UUIDPrimaryKeyMixin, TimestampMixin, SoftDeleteMixin):
    """
    Skills 注册表的数据库模型。

    既存储自定义 Skills（由运营人员注册），也存储 MCP 发现的已提升为 Skill 的工具。
    Qdrant 向量索引以 ``skill_id`` 为键，单独存储嵌入向量。
    """

    __tablename__ = "skills"

    # 唯一业务标识符（如 "skill-leave-query"）
    skill_id: Mapped[str] = mapped_column(
        String(128), unique=True, nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(256), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    category: Mapped[str] = mapped_column(String(64), nullable=False, index=True)

    # 标签 JSON 数组
    tags: Mapped[list] = mapped_column(JSONB, default=list)

    # Skill 接受的参数的 JSON Schema 描述
    parameters: Mapped[dict] = mapped_column(JSONB, default=dict)

    # 调用此 Skill 所需的权限字符串列表
    required_permissions: Mapped[list] = mapped_column(JSONB, default=list)

    # 处理器引用："mcp:{server}:{tool}" 或 "builtin:{name}" 或 "custom:{module}.{func}"
    handler: Mapped[str] = mapped_column(String(512), nullable=False)

    # 执行超时秒数
    timeout: Mapped[int] = mapped_column(Integer, default=30)

    # 版本号
    version: Mapped[str] = mapped_column(String(32), default="1.0.0")

    # 状态：active | inactive | deprecated
    status: Mapped[str] = mapped_column(String(32), default="active", index=True)

    # 来源：custom | mcp | builtin
    source: Mapped[str] = mapped_column(String(32), default="custom")

    # 排序优先级权重（越大越重要）
    priority: Mapped[float] = mapped_column(Float, default=1.0)

    # 此 Skill 执行前是否需要 HITL 审批
    requires_approval: Mapped[bool] = mapped_column(Boolean, default=False)

    # 若 source == "mcp" 则为 MCP 服务器名称
    mcp_server: Mapped[str | None] = mapped_column(String(128), nullable=True)

    # 使用统计（由 SkillRanker 更新）
    call_count: Mapped[int] = mapped_column(Integer, default=0)
    last_called_at: Mapped[str | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    def to_dict(self) -> dict:
        """返回可序列化的字典表示。"""
        return {
            "id": self.id,
            "skill_id": self.skill_id,
            "name": self.name,
            "description": self.description,
            "category": self.category,
            "tags": self.tags,
            "parameters": self.parameters,
            "required_permissions": self.required_permissions,
            "handler": self.handler,
            "timeout": self.timeout,
            "version": self.version,
            "status": self.status,
            "source": self.source,
            "priority": self.priority,
            "requires_approval": self.requires_approval,
            "mcp_server": self.mcp_server,
            "call_count": self.call_count,
            "last_called_at": self.last_called_at.isoformat()
            if self.last_called_at
            else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }
